src/test/java/com/google/neighborgood/UserInfoServletTest.java

```diff
@@ -0,0 +1,422 @@
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+
+    // The POST request should lead to an error handling clause, there should be no entity in the
+    // datastore
+    assertEquals(0, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+
+    // Now we give the servlet an invalid phone number input
+    when(request.getParameter("address-input")).thenReturn("4xxx Centre Avenue");
+    when(request.getParameter("phone-input")).thenReturn("");
+
+    try {
+      new UserInfoServlet().doPost(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+    assertEquals(0, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+
+    // Finally we give the servlet a valid phone number input
+    when(request.getParameter("phone-input")).thenReturn("4xxxxxxxxx");
+
+    try {
+      new UserInfoServlet().doPost(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+
+    assertEquals(1, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+    PreparedQuery results = ds.prepare(new Query("UserInfo"));
+    Entity entity = results.asSingleEntity();
+    if (entity == null) {
+      // The entity can't be null
+      assertTrue(false);
+    }
+
+    // Test the stored personal information
+    assertEquals("Leonard", (String) entity.getProperty("nickname"));
+    assertEquals("4xxx Centre Avenue", (String) entity.getProperty("address"));
+    assertEquals("4xxxxxxxxx", (String) entity.getProperty("phone"));
+    assertEquals("leonardzhang@google.com", (String) entity.getProperty("email"));
+    assertEquals("1234567890", (String) entity.getProperty("userId"));
+    assertEquals(0, (long) entity.getProperty("points"));
+  }
+
+  @Test
+  public void doGetTest() throws IOException {
+    HttpServletRequest request = mock(HttpServletRequest.class);
+    HttpServletResponse response = mock(HttpServletResponse.class);
+
+    // Check whether the datastore is empty before the test
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    assertEquals(0, ds.prepare(new Query("UserInfo")).countEntities(withLimit(10)));
+
+    Entity dummy = new Entity("UserInfo");
+    dummy.setProperty("nickname", "Leonard");
+    dummy.setProperty("address", "xxx");
+    dummy.setProperty("phone", "xxx");
+    dummy.setProperty("email", "test@example.com");
+    dummy.setProperty("userId", "1234567890");
+    dummy.setProperty("points", 0);
+    ds.put(dummy);
+
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    try {
+      new UserInfoServlet().doGet(request, response);
+    } catch (IOException e) {
+      assertTrue(false);
+    }
+
+    // After sending the GET request, the doGet function should output the json string
+    writer.flush();
+    assertTrue(stringWriter.toString().contains("[\"Leonard\",\"xxx\",\"xxx\"]"));
```

Comment:
It would be good to test the content type header of the response object, as well, to ensure that the server tells the client that it is returning JSON.

Reply:
Sure. I tried multiple ways to get the content type header of the response object, but since it's a mock object, the content type header is always null. So the only thing I can do is to use Mockito to verify that the servlet file has tried to set its content type to `application/json`.